Mathlib/Util/CaptureException.lean

```diff
@@ -0,0 +1,52 @@
+/-
+Copyright (c) 2024 Matthew Robert Ballard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matthew Robert Ballard, Damiano Testa
+-/
+
+import Lean.Elab.Command
+
+/-!
+# `#guard_exceptions` -- an analogue of `#guard_msgs` for parsers
+-/
+
+namespace Mathlib.GuardExceptions
+
+open Lean Parser Elab Command
+/-- `captureException env s input` uses the given `Environment` `env` to parse the `String` `input`
+using the `ParserFn` `s`. -/
+def captureException (env : Environment) (s : ParserFn) (input : String) : Except String Syntax :=
+  let ictx := mkInputContext input "<input>"
+  let s := s.run ictx { env, options := {} } (getTokenTable env) (mkParserState input)
+  if !s.allErrors.isEmpty then
+    .error (s.toErrorMsg ictx)
+  else if ictx.input.atEnd s.pos then
+    .ok s.stxStack.back
+  else
+    .error ((s.mkError "end of input").toErrorMsg ictx)
+
+/-- `#guard_exceptions in` is based on `#guard_msgs in`, except that it captures parsing exceptions.
```

Comment:
I propose that we redesign this a bit. Let's have a `#parse` command that tests a ParserFn:
```
"#parse " ident " => " str
```
Then, in test files, you would write `#guard_msgs in #parse Mathlib.Stacks.stacksTagFn => "A05"`

Yes, it's a bit wordier, but (1) it's more composable and (2) I think this is more likely for someone to be able to guess roughly what this is for when they see it.

Reply:
I like this suggestion very much, thanks!  I implemented what you are saying (I hope!) and it looks great!